### PR TITLE
sql: support unnest(listany)

### LINKED ISSUE #3800: sql: support unnest function
```
SELECT unnest(list[1, 2, 3])
1
2
3
```

### PR DESCRIPTION
**Pull request by sploiselle**
Closes #3800

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4639)

### CODE REVIEW

File: test/sqllogictest/list.slt [hunk 1]
```diff
@@ -1086,6 +1086,35 @@ SELECT list_length_max((LIST[1]::INT LIST), -1)
 query error invalid dimension: -1; must use value within \[1, 1\]
 SELECT list_length_max((LIST[1]::INT LIST), LIST[-1][1])
 
+# 🔬🔬 unnest
+
+query I
+SELECT * FROM unnest(LIST[1,2,3])
```
Comment (JLDLaughlin) [inline test/sqllogictest/list.slt:1092]: So in `psql` I see that I can execute both:
```
jessicalaughlin=# select * from unnest(array[1,2]);
 unnest
--------
      1
      2
(2 rows)
```
and
```
jessicalaughlin=# select unnest(array[1,2]);
 unnest
--------
      1
      2
(2 rows)
```

Is this true of our table functions? If so, could you add a test for the second case?

**Review by JLDLaughlin — APPROVED** [review]
looks good to me!

**Review by benesch — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4639#-:-MKfgpZS-1uYR78cOflC:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @JLDLaughlin and @sploiselle)

___
*[test/sqllogictest/list.slt, line 1092 at r1](https://reviewable.io/reviews/materializeinc/materialize/4639#-MKezOrw23zU4-GiWLvG-r1-1092:-MKfgZII6WcG8mnr7eUx:b-2u7e5w) ([raw file](https://github.com/materializeinc/materialize/blob/5fc558d798d8f6aebd981a467596f88df8558628/test/sqllogictest/list.slt#L1092)):*
<details><summary><i>Previously, JLDLaughlin (Jessica Laughlin) wrote…</i></summary><blockquote>

So in `psql` I see that I can execute both:
```
jessicalaughlin=# select * from unnest(array[1,2]);
 unnest
--------
      1
      2
(2 rows)
```
and
```
jessicalaughlin=# select unnest(array[1,2]);
 unnest
--------
      1
      2
(2 rows)
```

Is this true of our table functions? If so, could you add a test for the second case? 
</blockquote></details>

It is not true! The semantics of table functions in scalar position are total nonsense in complicated cases, so we've resisted adding that to date.
___
*[test/sqllogictest/list.slt, line 1116 at r1](https://reviewable.io/reviews/materializeinc/materialize/4639#-MKfgfiK7a39liPJZCMJ:-MKfgfiK7a39liPJZCMK:b-pnhmnh) ([raw file](https://github.com/materializeinc/materialize/blob/5fc558d798d8f6aebd981a467596f88df8558628/test/sqllogictest/list.slt#L1116)):*
> ```
> 
> query error Cannot call function unnest\(unknown\): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts
> SELECT * FROM unnest(NULL)
> ```

Can you add a test the same test but without the error (`unnest(NULL::blah list)` I think) for the empty-if-null behavior?

**Review by sploiselle — COMMENTED** [review]
ty for the quick reviews, y'all

> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4639#-:-MKftE8H1zm5_8E8aqRD:b-9axskt)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @benesch and @JLDLaughlin)

___
*[test/sqllogictest/list.slt, line 1116 at r1](https://reviewable.io/reviews/materializeinc/materialize/4639#-MKfgfiK7a39liPJZCMJ:-MKftBMg8Qq6DaG5MMnn:bb6un1j) ([raw file](https://github.com/materializeinc/materialize/blob/5fc558d798d8f6aebd981a467596f88df8558628/test/sqllogictest/list.slt#L1116)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Can you add a test the same test but without the error (`unnest(NULL::blah list)` I think) for the empty-if-null behavior?
</blockquote></details>

Done

**Pull request merged**